README.md
# clj-paste-server

A simple Clojure web app that lets you
* Write your data by POSTing it as JSON to /store/path.
* Read that same data (and possibly more) by GETing /store/path.

## Usage

1. Run the project using lein.
```$ lein run```

2. By default, you can find the server at `https://localhost:3443`.

### Example

1. First, we POST some data
```
$ curl --insecure \
    -X POST https://localhost:3443/store/mypath \
    -H "Content-Type: application/json" \
    -d '{"data": "mydata"}'
```

2. Then, we GET the data
```
$ curl --insecure \
    -X GET https://localhost:3443/store/mypath
```

3. The body of the response looks something like this
```
[{"data":"mydata",
    "time":"2019-07-15T11:13:49.961410Z"}]
```

## License

Copyright © 2019 Joel Söderman

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2019-07-14
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2019-07-14
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/paste-server/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/paste-server/compare/0.1.0...0.1.1


test/paste_server/core_test.clj
(ns paste-server.core-test
  (:require
    [clojure.test :refer :all]
    [clj-http.client :as http]
    [cheshire.core :as json]
    [paste-server.core :refer :all]))

(defn post-get-body [path data]
  (let [json-data (json/generate-string {:debug true
                                         :data data})
        response (http/post path {:insecure? true
                                  :body json-data})]
    (json/parse-string (:body response) true)))

(deftest a-test
  (testing "POST to store."
    (is (= (post-get-body "https://localhost:3443/store/hej" "hejsan!")
           {:where "hej"
            :data "hejsan!"}))))


project.clj
(defproject paste-server "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :license {:name "EPL-2.0 OR GPL-2.0-or-later WITH Classpath-exception-2.0"
            :url "https://www.eclipse.org/legal/epl-2.0/"}

  :dependencies [[org.clojure/clojure "1.10.0"]
                 [ring/ring-defaults "0.3.2"]
                 [jumblerg/ring-cors "2.0.0"]
                 [environ "1.1.0"]
                 [hiccup "1.0.5"]
                 [bananaoomarang/ring-debug-logging "1.1.0"]
                 [clojure.java-time "0.3.2"]
                 [ring "1.7.1"]
                 [cheshire "5.8.1"]
                 [clj-http "3.10.0"]
                 [compojure "1.6.1"]]

  :plugins [[cider/cider-nrepl "0.22.0-beta8"]
            [lein-environ "1.1.0"]]

  :repl {:plugins [[cider/cider-nrepl "0.22.0-beta8"]]}
  :repl-options {:init-ns paste-server.core}

  :profiles {:dev {:main paste-server.core/-dev-main
                   :env {:port "3000"
                         :ssl-port "3443"
                         :keystore "cert/keystore.jks"
                         :password "password"
                         :domain "localhost:3443"}}}

  :main paste-server.core/-main)



src/paste_server/core.clj
(ns paste-server.core
  (:require ;; misc
            [java-time :as time]
            [cheshire.core :as json]
            [environ.core :as environ]

            ;; compojure
            [compojure.route]
            [compojure.core :refer [defroutes GET POST]]

            ;; ring
            [ring.adapter.jetty :as jetty]
            [ring.util.request :as request]
            [ring.util.response :as response]

            ;; ring middleware
            [ring.middleware.defaults :refer :all]
            [jumblerg.middleware.cors :refer [wrap-cors]]
            [ring.middleware.reload :refer [wrap-reload]]
            [ring-debug-logging.core :refer [wrap-with-logger]]

            ;; project
            [paste-server.views :as views]
            [paste-server.store :as store]))

;; ++++++++++
;; + Routes +
;; ++++++++++

(defn store-GET
  "Handles GET-request to /store/:where."
  [where]
  (json/generate-string
    (store/store-read where)))

(defn parse-json-body [req]
  (json/parse-string
   (request/body-string req)
   true))

(defn store-POST
  "Handles POST-request to /store/:where."
  [where req]
  (let [body (parse-json-body req)
        data (str (:data body ""))
        debug? (:debug body)]

    (do (store/store-write
          where
          {:data data
           :time (str (time/instant)) })

        (if-not debug?
          ""
          (json/generate-string
            {:where (str where)
             :data data})))))

(defroutes routes
  (GET "/" req
    (-> (response/response views/index)
        (response/content-type "text/html")))

  (GET "/store/:where" [where :as req]
    (-> (response/response (store-GET where))
        (response/content-type "application/json")))

  (POST "/store/:where" [where :as req]
    (-> (response/response (store-POST where req))
        (response/content-type "application/json")))

  (compojure.route/not-found
   (response/response "")))

;; +++++++
;; + App +
;; +++++++

(def app
  (-> #'routes
      (wrap-cors routes ".*")
      (wrap-defaults secure-api-defaults)))

(def options
  {:ssl? true
   :port (Integer. (environ/env :port))
   :ssl-port (Integer. (environ/env :ssl-port))
   :keystore (environ/env :keystore)
   :key-password (environ/env :password)})

(defn -dev-main []
  (jetty/run-jetty
    (-> #'app
        wrap-reload
        wrap-with-logger)

    options))

(defn -main [] (jetty/run-jetty app options))


src/paste_server/views.clj
(ns paste-server.views
  (:require [environ.core :as environ]
            [hiccup.core :refer :all]
            [clojure.string :refer [join]]))

(def domain (environ/env :domain))

(def index
  (html
    [:h1 "Welcome!"]
    [:p "Here you can"]
    [:ul [:li "Write your data by POSTing it as JSON to /store/path."]
         [:li "Read that same data (and possibly more) by GETing /store/path."]]
    [:h2 "Example"]
    [:ol
      [:li [:p "First, we POST some data"]
           [:pre (join "\n"
                      ["$ curl --insecure \\"
                       (str "    -X POST https://" domain "/store/mypath \\")
                       "    -H \"Content-Type: application/json\" \\"
                       "    -d '{\"data\": \"mydata\"}'"])]]

      [:li [:p "Then, we GET the data"]
           [:pre (join "\n"
                       ["$ curl --insecure \\"
                        (str "    -X GET https://" domain "/store/mypath")])]]

      [:li [:p "The body of the response looks something like this"]
           [:pre (join "\n"
                       ["[{\"data\":\"mydata\","
                        "  \"time\":\"2019-07-15T11:13:49.961410Z\"}]"])]]]))
